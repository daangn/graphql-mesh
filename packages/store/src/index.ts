import { promises as fsPromises } from 'fs';
import { join } from 'path';
import { pathExists, writeFile, flatString, jsonFlatStringify } from '@graphql-mesh/utils';

const { readFile } = fsPromises;

export class ReadonlyStoreError extends Error {}

export class ValidationError extends Error {
  constructor(message: string, public originalError: Error) {
    super(message);
  }
}

export type StoreStorageAdapter<TData = any, TKey = string> = {
  exists: (key: TKey) => Promise<boolean>;
  read: (key: TKey, options: ProxyOptions<TData>) => Promise<TData>;
  write: (key: TKey, data: TData, options: ProxyOptions<TData>) => Promise<TData>;
};

export class InMemoryStoreStorageAdapter implements StoreStorageAdapter {
  private data = new Map<string, any>();

  async exists(key: string): Promise<boolean> {
    return this.data.has(key);
  }

  async read<TData>(key: string, options: ProxyOptions<any>): Promise<TData> {
    return this.data.get(key);
  }

  async write<TData>(key: string, data: TData, options: ProxyOptions<any>): Promise<void> {
    this.data.set(key, data);
  }

  clear() {
    this.data.clear();
  }
}

export class FsStoreStorageAdapter implements StoreStorageAdapter {
  async exists(key: string): Promise<boolean> {
    return pathExists(key);
  }

  async read<TData>(key: string, options: ProxyOptions<any>): Promise<TData> {
    const asString = await readFile(key, 'utf-8');

    return options.parse(asString);
  }

  async write<TData>(key: string, data: TData, options: ProxyOptions<any>): Promise<void> {
    const asString = options.serialize(data);
    return writeFile(key, asString);
  }
}

export type StoreProxy<TData> = {
  set(value: TData): Promise<void>;
  get(): Promise<TData>;
};

export type ProxyOptions<TData> = {
  parse: (content: string) => TData;
  serialize: (value: TData) => string;
  validate: (oldValue: TData, newValue: TData) => void;
};

export type StoreFlags = {
  readonly: boolean;
  validate: boolean;
};

export enum PredefinedProxyOptionsName {
  JsonWithoutValidation = 'JsonWithoutValidation',
  StringWithoutValidation = 'StringWithoutValidation',
}

export const PredefinedProxyOptions: Record<PredefinedProxyOptionsName, ProxyOptions<any>> = {
  JsonWithoutValidation: {
    parse: v => JSON.parse(v),
    serialize: v => jsonFlatStringify(v),
    validate: () => null,
  },
  StringWithoutValidation: {
    parse: v => flatString(v),
    serialize: v => flatString(v),
    validate: () => null,
  },
};

export class MeshStore {
  constructor(public identifier: string, protected storage: StoreStorageAdapter, public flags: StoreFlags) {}

  child(childIdentifier: string): MeshStore {
    return new MeshStore(join(this.identifier, childIdentifier), this.storage, this.flags);
  }

  proxy<TData>(id: string, options: ProxyOptions<TData>): StoreProxy<TData> {
    const path = join(this.identifier, id);
    let value: TData | null | undefined;

    return {
      get: async () => {
        if (typeof value !== 'undefined') {
          return value;
        }

        if (!(await this.storage.exists(path))) {
          return undefined;
        }

        return this.storage.read(path, options);
      },
      set: async newValue => {
        if (this.flags.readonly) {
          throw new ReadonlyStoreError(
            `Unable to set value for "${id}" under "${this.identifier}" because the store is in read-only mode.`
          );
        }

        if (this.flags.validate) {
          if (value && newValue) {
            try {
              options.validate(value, newValue);
            } catch (e) {
              throw new ValidationError(`Validation failed for "${id}" under "${this.identifier}"!`, e);
            }
          }
        }

        value = newValue;
        await this.storage.write(path, value, options);
      },
    };
  }
}
