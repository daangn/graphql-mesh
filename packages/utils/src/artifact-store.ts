export type ArtifactStore<TKey = string> = {
  child(identifier: TKey): ArtifactStore;
  has(id: TKey): Promise<boolean>;
  get<TData>(id: TKey): Promise<TData>;
  path(id: TKey): string;
  set<TData>(id: TKey, data: TData): Promise<void>;
};

export class FsArtifactStore implements ArtifactStore<string> {
  child(identifier: string): ArtifactStore<string> {
    throw new Error('Method not implemented.');
  }

  has(id: string): Promise<boolean> {
    throw new Error('Method not implemented.');
  }

  get<TData>(id: string): Promise<TData> {
    throw new Error('Method not implemented.');
  }

  path(id: string): string {
    throw new Error('Method not implemented.');
  }

  set<TData>(id: string, data: TData): Promise<void> {
    throw new Error('Method not implemented.');
  }
}
