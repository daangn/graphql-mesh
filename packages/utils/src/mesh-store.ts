export type MeshStore<TKey = string> = {
  child(identifier: TKey): MeshStore;
  has(id: TKey): Promise<boolean>;
  get<TData>(id: TKey): Promise<TData>;
  path(id: TKey): string;
  set<TData>(id: TKey, data: TData): Promise<void>;
};

export class InMemoryMeshStore implements MeshStore {
  protected data = new Map<string, string>();

  child(identifier: string): MeshStore<string> {
    throw new Error('Method not implemented.');
  }

  async has(id: string): Promise<boolean> {
    return this.data.has(id);
  }

  async get<TData>(id: string): Promise<TData> {
    return (this.data.get(id) as any) as TData;
  }

  path(id: string): string {
    throw new Error('Method not implemented.');
  }

  set<TData>(id: string, data: TData): Promise<void> {
    throw new Error('Method not implemented.');
  }
}

export class FsMeshStore implements MeshStore {
  child(identifier: string): MeshStore<string> {
    throw new Error('Method not implemented.');
  }

  has(id: string): Promise<boolean> {
    throw new Error('Method not implemented.');
  }

  get<TData>(id: string): Promise<TData> {
    throw new Error('Method not implemented.');
  }

  path(id: string): string {
    throw new Error('Method not implemented.');
  }

  set<TData>(id: string, data: TData): Promise<void> {
    throw new Error('Method not implemented.');
  }
}
