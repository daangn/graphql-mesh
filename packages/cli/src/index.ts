import { findAndParseConfig } from '@graphql-mesh/config';
import { getMesh } from '@graphql-mesh/runtime';
import * as yargs from 'yargs';
import { generateTsArtifacts } from './commands/ts-artifacts';
import { serveMesh } from './commands/serve/serve';
import { isAbsolute, resolve, join } from 'path';
import { existsSync } from 'fs';
import { logger } from './logger';
import { FsStoreStorageAdapter, MeshStore, StoreFlags } from '@graphql-mesh/store';
import { dumpSchema } from './commands/dump-schema';

export { generateTsArtifacts, serveMesh };

export async function graphqlMesh() {
  let baseDir = process.cwd();

  return yargs
    .help()
    .option('r', {
      alias: 'require',
      describe: 'Loads specific require.extensions before running the codegen and reading the configuration',
      type: 'array' as const,
      default: [],
      coerce: (externalModules: string[]) =>
        Promise.all(
          externalModules.map(module => {
            const localModulePath = resolve(process.cwd(), module);
            const islocalModule = existsSync(localModulePath);
            return import(islocalModule ? localModulePath : module);
          })
        ),
    })
    .option('dir', {
      describe: 'Modified the base directory to use for looking for meshrc config file',
      type: 'string',
      default: process.cwd(),
      coerce: dir => {
        if (isAbsolute(dir)) {
          baseDir = dir;
        } else {
          baseDir = resolve(process.cwd(), dir);
        }
      },
    })
    .command<{ port: number; prod: boolean; validate: boolean }>(
      'serve',
      'Serves a GraphQL server with GraphQL interface to test your Mesh API',
      builder => {
        builder.option('port', {
          type: 'number',
        });
        builder.option('prod', {
          type: 'boolean',
          coerce: value => value || process.env.NODE_ENV?.toLowerCase() === 'production',
        });
      },
      async args => {
        try {
          await serveMesh({
            baseDir,
            argsPort: args.port,
            store: new MeshStore(join(baseDir, '.mesh'), new FsStoreStorageAdapter(), {
              readonly: args.prod,
              validate: false,
            }),
          });
        } catch (e) {
          logger.error('Unable to serve mesh: ', e);
        }
      }
    )
    .command(
      'validate',
      'Validates artifacts',
      builder => {},
      async args => {
        await buildArtifacts(baseDir, {
          readonly: false, // This will be true when handlers are ready
          validate: true,
        });
      }
    )
    .command(
      'build',
      'Builds artifacts',
      builder => {},
      async args => {
        await buildArtifacts(baseDir, {
          readonly: false,
          validate: false,
        });
      }
    ).argv;
}

const buildArtifacts = async (baseDir: string, flags: StoreFlags) => {
  const store = new MeshStore(join(baseDir, '.mesh'), new FsStoreStorageAdapter(), flags);
  const meshConfig = await findAndParseConfig({
    dir: baseDir,
    ignoreAdditionalResolvers: true,
    store,
  });
  const { schema, destroy, rawSources } = await getMesh(meshConfig);

  await Promise.all([
    generateTsArtifacts(schema, rawSources, meshConfig.mergerType, meshConfig.documents, false, store),
    dumpSchema(schema, baseDir, store),
  ]);
  destroy();
};
