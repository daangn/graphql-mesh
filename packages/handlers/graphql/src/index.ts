import {
  GetMeshSourceOptions,
  MeshHandler,
  MeshSource,
  ResolverData,
  YamlConfig,
  KeyValueCache,
} from '@graphql-mesh/types';
import { UrlLoader } from '@graphql-tools/url-loader';
import {
  GraphQLSchema,
  buildClientSchema,
  introspectionFromSchema,
  buildSchema,
  IntrospectionQuery,
  parse,
  execute,
  GraphQLResolveInfo,
  DocumentNode,
  Kind,
  buildASTSchema,
} from 'graphql';
import { introspectSchema } from '@graphql-tools/wrap';
import {
  getInterpolatedHeadersFactory,
  ResolverDataBasedFactory,
  getHeadersObject,
  loadFromModuleExportExpression,
  getInterpolatedStringFactory,
  getCachedFetch,
  readFileOrUrlWithCache,
} from '@graphql-mesh/utils';
import { ExecutionParams, AsyncExecutor } from '@graphql-tools/delegate';
import { PredefinedProxyOptions, StoreProxy } from '@graphql-mesh/store';

const APOLLO_GET_SERVICE_DEFINITION_QUERY = /* GraphQL */ `
  query __ApolloGetServiceDefinition__ {
    _service {
      sdl
    }
  }
`;

export default class GraphQLHandler implements MeshHandler {
  private config: YamlConfig.GraphQLHandler;
  private baseDir: string;
  private cache: KeyValueCache<any>;
  private introspection: StoreProxy<IntrospectionQuery>;
  private sdl: StoreProxy<string>;

  constructor({ config, baseDir, cache, store }: GetMeshSourceOptions<YamlConfig.GraphQLHandler>) {
    this.config = config;
    this.baseDir = baseDir;
    this.cache = cache;
    this.introspection = store.proxy<IntrospectionQuery>('schema.json', PredefinedProxyOptions.JsonWithoutValidation);
    this.sdl = store.proxy<string>('schema.graphql', PredefinedProxyOptions.StringWithoutValidation);
  }

  async getMeshSource(): Promise<MeshSource> {
    const { endpoint, schemaHeaders: configHeaders, introspection } = this.config;
    const customFetch = getCachedFetch(this.cache);

    if (endpoint.endsWith('.js') || endpoint.endsWith('.ts')) {
      // Loaders logic should be here somehow
      const schemaOrStringOrDocumentNode = await loadFromModuleExportExpression<GraphQLSchema | string | DocumentNode>(
        endpoint,
        { cwd: this.baseDir }
      );
      let schema: GraphQLSchema;
      if (schemaOrStringOrDocumentNode instanceof GraphQLSchema) {
        schema = schemaOrStringOrDocumentNode;
      } else if (typeof schemaOrStringOrDocumentNode === 'string') {
        schema = buildSchema(schemaOrStringOrDocumentNode);
      } else if (
        typeof schemaOrStringOrDocumentNode === 'object' &&
        schemaOrStringOrDocumentNode?.kind === Kind.DOCUMENT
      ) {
        schema = buildASTSchema(schemaOrStringOrDocumentNode);
      } else {
        throw new Error(
          `Provided file '${endpoint} exports an unknown type: ${typeof schemaOrStringOrDocumentNode}': expected GraphQLSchema, SDL or DocumentNode.`
        );
      }
      return {
        schema,
      };
    } else if (endpoint.endsWith('.graphql')) {
      const rawSDL = await readFileOrUrlWithCache<string>(endpoint, this.cache, {
        cwd: this.baseDir,
        allowUnknownExtensions: true,
      });
      const schema = buildSchema(rawSDL);
      return {
        schema,
      };
    }
    const urlLoader = new UrlLoader();
    const getExecutorAndSubscriberForParams = (
      params: ExecutionParams,
      headersFactory: ResolverDataBasedFactory<Headers>,
      endpointFactory: ResolverDataBasedFactory<string>
    ) => {
      const resolverData: ResolverData = {
        root: {},
        args: params.variables,
        context: params.context,
      };
      const headers = getHeadersObject(headersFactory(resolverData));
      const endpoint = endpointFactory(resolverData);
      return urlLoader.getExecutorAndSubscriberAsync(endpoint, {
        customFetch,
        ...this.config,
        headers,
      });
    };
    let schemaHeaders =
      typeof configHeaders === 'string'
        ? await loadFromModuleExportExpression(configHeaders, { cwd: this.baseDir })
        : configHeaders;
    if (typeof schemaHeaders === 'function') {
      schemaHeaders = schemaHeaders();
    }
    if (schemaHeaders && 'then' in schemaHeaders) {
      schemaHeaders = await schemaHeaders;
    }
    const schemaHeadersFactory = getInterpolatedHeadersFactory(schemaHeaders || {});
    const introspectionExecutor: AsyncExecutor = async (params): Promise<any> => {
      const { executor } = await getExecutorAndSubscriberForParams(params, schemaHeadersFactory, () => endpoint);
      return executor(params);
    };
    let introspectionResult = await this.introspection.get();
    if (!introspectionResult) {
      if (introspection) {
        const result = await urlLoader.handleSDLAsync(introspection, {
          customFetch,
          ...this.config,
          headers: schemaHeaders,
        });
        introspectionResult = introspectionFromSchema(result.schema);
      } else {
        introspectionResult = introspectionFromSchema(await introspectSchema(introspectionExecutor));
      }
      await this.introspection.set(introspectionResult);
    }
    const nonExecutableSchema = buildClientSchema(introspectionResult);
    const operationHeadersFactory = getInterpolatedHeadersFactory(this.config.operationHeaders);
    const endpointFactory = getInterpolatedStringFactory(endpoint);
    const queryType = nonExecutableSchema.getQueryType();
    const queryTypeFieldMap = queryType.getFields();
    if ('_service' in queryTypeFieldMap) {
      const _serviceField = queryTypeFieldMap._service;
      if ('resolve' in _serviceField) {
        _serviceField.resolve = async () => {
          let sdl = await this.sdl.get();
          if (!sdl) {
            const sdlQueryResult = await introspectionExecutor({
              document: parse(APOLLO_GET_SERVICE_DEFINITION_QUERY),
            });
            sdl = sdlQueryResult?.data?._service?.sdl;
            await this.sdl.set(sdl);
          }
          return {
            sdl,
          };
        };
      }
    }
    const isSdlQuery = (info: GraphQLResolveInfo) =>
      info.fieldName === '_service' &&
      info.fieldNodes[0].selectionSet.selections[0].kind === 'Field' &&
      info.fieldNodes[0].selectionSet.selections[0].name.value === 'sdl';
    return {
      schema: nonExecutableSchema,
      executor: async params => {
        if (isSdlQuery(params.info)) {
          return execute(nonExecutableSchema, params.document);
        }
        const { executor } = await getExecutorAndSubscriberForParams(params, operationHeadersFactory, endpointFactory);
        return executor(params) as any;
      },
      subscriber: async params => {
        const { subscriber } = await getExecutorAndSubscriberForParams(
          params,
          operationHeadersFactory,
          endpointFactory
        );
        return subscriber(params);
      },
      batch: 'batch' in this.config ? this.config.batch : true,
    };
  }
}
