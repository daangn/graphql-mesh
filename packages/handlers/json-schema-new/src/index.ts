import {
  EnumTypeComposerValueConfigDefinition,
  GraphQLJSON,
  isSomeInputTypeComposer,
  SchemaComposer,
} from 'graphql-compose';
import { JSONSchema } from '@json-schema-tools/meta-schema';
import traverse from '@json-schema-tools/traverse';
import { GraphQLBoolean, GraphQLFloat, GraphQLInt, GraphQLString } from 'graphql';
import {
  GraphQLBigInt,
  GraphQLDate,
  GraphQLDateTime,
  GraphQLEmailAddress,
  GraphQLIPv4,
  GraphQLIPv6,
  GraphQLTime,
  GraphQLURL,
  GraphQLVoid,
  RegularExpression,
} from 'graphql-scalars';

export function visitJSONSchema(schema: JSONSchema, schemaComposer: SchemaComposer, prefix: string) {
  const refTypeComposerMap = new Map<any, any>();

  function ensureTypeComposer(maybeTypeComposer: any) {
    return refTypeComposerMap.get(maybeTypeComposer) || maybeTypeComposer;
  }

  return traverse(schema, (subSchema, _, path) => {
    const getTypeComposer = () => {
      if (typeof subSchema === 'boolean') {
        return subSchema ? schemaComposer.getAnyTC(GraphQLJSON) : undefined;
      }
      subSchema.title = subSchema.title || path.split('/').join('_');
      subSchema.title = prefix + subSchema.title;
      if (subSchema.pattern) {
        return schemaComposer.createScalarTC(
          new RegularExpression(subSchema.title, new RegExp(subSchema.pattern), {
            description: subSchema.description,
          })
        );
      }
      if (subSchema.const) {
        return schemaComposer.createScalarTC(
          new RegularExpression(subSchema.title, new RegExp(subSchema.const), {
            description: subSchema.description,
          })
        );
      }
      if (subSchema.enum) {
        const values: Record<string, EnumTypeComposerValueConfigDefinition> = {};
        for (const value of subSchema.enum) {
          values[value] = subSchema.enum;
        }
        return schemaComposer.createEnumTC({
          name: subSchema.title,
          values,
        });
      }
      if (subSchema.allOf) {
        const typeComposers = subSchema.allOf.map(ensureTypeComposer);
        // Scalars cannot be in union type
        if (typeComposers.some(typeComposer => isSomeInputTypeComposer(typeComposer))) {
          return schemaComposer.getAnyTC(GraphQLJSON);
        }
        return schemaComposer.createUnionTC({
          name: subSchema.title,
          description: subSchema.description,
          types: typeComposers,
        });
      }
      if (subSchema.anyOf) {
        const typeComposers = subSchema.anyOf.map(ensureTypeComposer);
        // Scalars cannot be in union type
        if (typeComposers.some(typeComposer => isSomeInputTypeComposer(typeComposer))) {
          return schemaComposer.getAnyTC(GraphQLJSON);
        }
        return schemaComposer.createUnionTC({
          name: subSchema.title,
          description: subSchema.description,
          types: typeComposers,
        });
      }
      if (subSchema.oneOf) {
        const typeComposers = subSchema.oneOf.map(ensureTypeComposer);
        // Scalars cannot be in union type
        if (typeComposers.some(typeComposer => isSomeInputTypeComposer(typeComposer))) {
          return schemaComposer.getAnyTC(GraphQLJSON);
        }
        return schemaComposer.createUnionTC({
          name: subSchema.title,
          description: subSchema.description,
          types: typeComposers,
        });
      }
      switch (subSchema.type) {
        case 'boolean':
          return schemaComposer.getAnyTC(GraphQLBoolean);
        case 'null':
          return schemaComposer.getAnyTC(GraphQLVoid);
        case 'integer':
          if (subSchema.format === 'int64') {
            return schemaComposer.getAnyTC(GraphQLBigInt);
          }
          return schemaComposer.getAnyTC(GraphQLInt);
        case 'number':
          return schemaComposer.getAnyTC(GraphQLFloat);
        case 'string':
          if (subSchema.minLength || subSchema.maxLength) {
            const coerceString = (v: any) => {
              if (v != null) {
                const vStr = v.toString();
                if (typeof subSchema.minLength !== 'undefined' && vStr.length > subSchema.minLength) {
                  throw new Error(`${subSchema.title} cannot be less than ${subSchema.minLength}`);
                }
                if (typeof subSchema.maxLength !== 'undefined' && vStr.length > subSchema.maxLength) {
                  throw new Error(`${subSchema.title} cannot be more than ${subSchema.maxLength}`);
                }
                return vStr;
              }
            };
            return schemaComposer.createScalarTC({
              name: subSchema.title,
              description: subSchema.description,
              serialize: coerceString,
              parseLiteral: coerceString,
              parseValue: ast => ast?.value && coerceString(ast.value),
            });
          }
          switch (subSchema.format) {
            case 'date-time':
              return schemaComposer.getAnyTC(GraphQLDateTime);
            case 'time':
              return schemaComposer.getAnyTC(GraphQLTime);
            case 'date':
              return schemaComposer.getAnyTC(GraphQLDate);
            case 'email':
            case 'idn-email':
              return schemaComposer.getAnyTC(GraphQLEmailAddress);
            case 'hostname':
              return schemaComposer.getAnyTC(GraphQLString);
            case 'ipv4':
              return schemaComposer.getAnyTC(GraphQLIPv4);
            case 'ipv6':
              return schemaComposer.getAnyTC(GraphQLIPv6);
            case 'uri':
            case 'uri-reference':
            case 'iri':
            case 'iri-reference':
            case 'uri-template':
              return schemaComposer.getAnyTC(GraphQLURL);
            case 'json-pointer':
              return schemaComposer.getAnyTC(GraphQLString);
            case 'relative-json-pointer':
              return schemaComposer.getAnyTC(GraphQLString);
            case 'regex':
              return schemaComposer.getAnyTC(GraphQLString);
            default:
              return schemaComposer.getAnyTC(GraphQLString);
          }
        case 'array':
          if (typeof subSchema.items === 'object' && !Array.isArray(subSchema.items)) {
            const typeComposer = ensureTypeComposer(subSchema.items);
            return typeComposer.getTypePlural();
          }
          if (subSchema.contains) {
            // Scalars cannot be in union type
            return schemaComposer.getAnyTC(GraphQLJSON).getTypePlural();
          }
          if (typeof subSchema.items === 'object' && Array.isArray(subSchema.items)) {
            const existingItems = [...(subSchema.items as any)];
            if (subSchema.additionalItems) {
              existingItems.push(subSchema.additionalItems);
            }
            const typeComposers = existingItems.map(ensureTypeComposer);
            if (typeComposers.some(existingItem => isSomeInputTypeComposer(existingItem))) {
              return schemaComposer.getAnyTC(GraphQLJSON);
            }
            const unionComposer = schemaComposer.createUnionTC({
              name: subSchema.title,
              description: subSchema.description,
              types: typeComposers,
            });
            return unionComposer.getTypePlural();
          }
        // eslint-disable-next-line no-fallthrough
        case 'object':
          // eslint-disable-next-line no-case-declarations
          const fieldMap: any = {};
          for (const propertyName in subSchema.properties) {
            const typeComposer = ensureTypeComposer(subSchema.properties[propertyName]);
            fieldMap[propertyName] = {
              type: subSchema.required?.includes(propertyName) ? typeComposer.getTypeNonNull() : typeComposer,
            };
          }
          if (subSchema.additionalProperties) {
            fieldMap.additionalProperties = {
              type: ensureTypeComposer(subSchema.additionalProperties),
              resolve: (root: any) => root,
            };
          }
          return schemaComposer.createObjectTC({
            name: subSchema.title,
            description: subSchema.description,
            fields: fieldMap,
          });
      }
    };
    const result = getTypeComposer();
    refTypeComposerMap.set(subSchema, result);
    return result;
  });
}
