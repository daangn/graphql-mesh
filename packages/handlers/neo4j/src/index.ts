import { isAbsolute } from 'path';
import { makeAugmentedSchema, inferSchema } from 'neo4j-graphql-js';
import neo4j, { Driver } from 'neo4j-driver';
import { YamlConfig, MeshHandler, GetMeshSourceOptions, MeshPubSub } from '@graphql-mesh/types';
import { loadTypedefs } from '@graphql-tools/load';
import { GraphQLFileLoader } from '@graphql-tools/graphql-file-loader';
import { CodeFileLoader } from '@graphql-tools/code-file-loader';
import { mergeTypeDefs } from '@graphql-tools/merge';
import { DocumentNode, parse } from 'graphql';
import { PredefinedProxyOptions, StoreProxy } from '@graphql-mesh/store';

export default class Neo4JHandler implements MeshHandler {
  private config: YamlConfig.Neo4JHandler;
  private baseDir: string;
  private pubsub: MeshPubSub;
  private typeDefs: StoreProxy<DocumentNode>;

  constructor({ config, baseDir, pubsub, store }: GetMeshSourceOptions<YamlConfig.Neo4JHandler>) {
    this.config = config;
    this.baseDir = baseDir;
    this.pubsub = pubsub;
    this.typeDefs = store.proxy('typeDefs.json', PredefinedProxyOptions.JsonWithoutValidation);
  }

  private driver: Driver;

  getDriver() {
    if (!this.driver) {
      this.driver = neo4j.driver(this.config.url, neo4j.auth.basic(this.config.username, this.config.password));
      this.pubsub.subscribe('destroy', () => this.driver.close());
    }
    return this.driver;
  }

  async getCachedTypeDefs() {
    let typeDefs = await this.typeDefs.get();
    if (!typeDefs) {
      if (this.config.typeDefs) {
        const typeDefsArr = await loadTypedefs(this.config.typeDefs, {
          cwd: isAbsolute(this.config.typeDefs) ? null : this.baseDir,
          loaders: [new CodeFileLoader(), new GraphQLFileLoader()],
          assumeValid: true,
          assumeValidSDL: true,
        });
        typeDefs = mergeTypeDefs(typeDefsArr.map(source => source.document));
      } else {
        typeDefs = await inferSchema(this.getDriver(), {
          alwaysIncludeRelationships: this.config.alwaysIncludeRelationships,
        });
        if (typeof typeDefs === 'string') {
          typeDefs = parse(typeDefs);
        }
      }
      await this.typeDefs.set(typeDefs);
    }
    return typeDefs;
  }

  async getMeshSource() {
    const typeDefs: DocumentNode | string = await this.getCachedTypeDefs();

    const schema = makeAugmentedSchema({ typeDefs, config: { experimental: true } });

    return {
      schema,
      contextBuilder: async () => ({ driver: this.getDriver(), neo4jDatabase: this.config.database }),
    };
  }
}
