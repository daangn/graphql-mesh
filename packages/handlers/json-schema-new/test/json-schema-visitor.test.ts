import { ObjectTypeComposer, SchemaComposer } from 'graphql-compose';
import { visitJSONSchema } from '../src';
import $RefParser from '@apidevtools/json-schema-ref-parser';
import titleizer from '@json-schema-tools/titleizer';

describe('visitor', () => {
  it('should do something', async () => {
    const schemaComposer = new SchemaComposer();
    const someSchema: any = {
      definitions: {
        app: {
          definitions: {
            domains: {
              items: {
                $ref: '#/definitions/domain',
              },
              type: 'array',
            },
            name: {
              pattern: '^[a-z][a-z0-9-]{2,30}$',
              type: 'string',
            },
          },
          properties: {
            domains: {
              $ref: '#/definitions/app/definitions/domains',
            },
            name: {
              $ref: '#/definitions/app/definitions/name',
            },
          },
          required: ['name'],
          type: 'object',
        },
        domain: {
          definitions: {
            name: {
              format: 'hostname',
              type: 'string',
            },
          },
          properties: {
            name: {
              $ref: '#/definitions/domain/definitions/name',
            },
          },
          required: ['name'],
          type: 'object',
        },
      },
      properties: {
        app: {
          $ref: '#/definitions/app',
        },
        domain: {
          $ref: '#/definitions/domain',
        },
      },
      type: 'object',
    };
    const bundled = await $RefParser.bundle(someSchema);
    const dereferenced = await $RefParser.dereference(bundled);
    expect(dereferenced).toMatchSnapshot();
    const typeComposer: ObjectTypeComposer<any> = visitJSONSchema(
      dereferenced as any,
      schemaComposer,
      'DomainAPP'
    ) as any;
    schemaComposer.Query.addFields({
      test: {
        type: typeComposer,
      },
    });
    expect(schemaComposer.toSDL()).toMatchSnapshot();
  });
});
