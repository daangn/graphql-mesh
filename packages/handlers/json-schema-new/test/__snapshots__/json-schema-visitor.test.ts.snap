// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`visitor should do something 1`] = `
Object {
  "definitions": Object {
    "app": Object {
      "definitions": Object {
        "domains": Object {
          "items": Object {
            "definitions": Object {
              "name": Object {
                "format": "hostname",
                "type": "string",
              },
            },
            "properties": Object {
              "name": Object {
                "format": "hostname",
                "type": "string",
              },
            },
            "required": Array [
              "name",
            ],
            "type": "object",
          },
          "type": "array",
        },
        "name": Object {
          "pattern": "^[a-z][a-z0-9-]{2,30}$",
          "type": "string",
        },
      },
      "properties": Object {
        "domains": Object {
          "items": Object {
            "definitions": Object {
              "name": Object {
                "format": "hostname",
                "type": "string",
              },
            },
            "properties": Object {
              "name": Object {
                "format": "hostname",
                "type": "string",
              },
            },
            "required": Array [
              "name",
            ],
            "type": "object",
          },
          "type": "array",
        },
        "name": Object {
          "pattern": "^[a-z][a-z0-9-]{2,30}$",
          "type": "string",
        },
      },
      "required": Array [
        "name",
      ],
      "type": "object",
    },
    "domain": Object {
      "definitions": Object {
        "name": Object {
          "format": "hostname",
          "type": "string",
        },
      },
      "properties": Object {
        "name": Object {
          "format": "hostname",
          "type": "string",
        },
      },
      "required": Array [
        "name",
      ],
      "type": "object",
    },
  },
  "properties": Object {
    "app": Object {
      "definitions": Object {
        "domains": Object {
          "items": Object {
            "definitions": Object {
              "name": Object {
                "format": "hostname",
                "type": "string",
              },
            },
            "properties": Object {
              "name": Object {
                "format": "hostname",
                "type": "string",
              },
            },
            "required": Array [
              "name",
            ],
            "type": "object",
          },
          "type": "array",
        },
        "name": Object {
          "pattern": "^[a-z][a-z0-9-]{2,30}$",
          "type": "string",
        },
      },
      "properties": Object {
        "domains": Object {
          "items": Object {
            "definitions": Object {
              "name": Object {
                "format": "hostname",
                "type": "string",
              },
            },
            "properties": Object {
              "name": Object {
                "format": "hostname",
                "type": "string",
              },
            },
            "required": Array [
              "name",
            ],
            "type": "object",
          },
          "type": "array",
        },
        "name": Object {
          "pattern": "^[a-z][a-z0-9-]{2,30}$",
          "type": "string",
        },
      },
      "required": Array [
        "name",
      ],
      "type": "object",
    },
    "domain": Object {
      "definitions": Object {
        "name": Object {
          "format": "hostname",
          "type": "string",
        },
      },
      "properties": Object {
        "name": Object {
          "format": "hostname",
          "type": "string",
        },
      },
      "required": Array [
        "name",
      ],
      "type": "object",
    },
  },
  "type": "object",
}
`;

exports[`visitor should do something 2`] = `
"type Query {
  test: DomainAPP
}

\\"\\"\\"
A field whose value matches the provided regular expression /^[a-z][a-z0-9-]{2,30}$/.
\\"\\"\\"
scalar DomainAPP_properties_app_properties_name

\\"\\"\\"
The \`String\` scalar type represents textual data, represented as UTF-8 character sequences. The String type is most often used by GraphQL to represent free-form human-readable text.
\\"\\"\\"
scalar String

type DomainAPP {
  app: DomainAPP_properties_app
  domain: DomainAPP_properties_app_properties_domains
}

type DomainAPP_properties_app {
  domains: [DomainAPP_properties_app_properties_domains]
  name: DomainAPP_properties_app_properties_name!
}

type DomainAPP_properties_app_properties_domains {
  name: String!
}"
`;
